### ISSUE TITLE
Hard to return preformatted json with application/json content-type

### ISSUE THREAD

**Issue by iabdulin** [ruby-grape/grape#1989]
For example, we have the following endpoint, returning preformatted json (I return json right from the database)
```ruby
get 'all_json' do
   {key: 123}.to_json
end
```
Here's what will be returned:
`"{\"key\":123}"` with `application/json` content-type
Grape applies additional `.to_json` call.

Now let's try to fix this problem, taking into assumption that I already have a prebuilt JSON in my endpoint.

Options:
1) Adding `env['api.format'] = :binary` or `:txt` — `{"key":123}` but empty content-type

2) Adding `content_type 'application/json'` or `header 'Content-Type', 'application/json'` to :txt or :binary format — `"{\"key\":123}"`

The only way I was able to solve this is by overwriting :json formatter with:

```ruby
formatter :json, ->(object, env) {
      if object.is_a?(String)
        object
      else
        return object.to_json if object.respond_to?(:to_json)
        Grape::Json.dump(object)
      end
    }
```

Maybe there's another way?

**Comment by dblock** [comment]
I think in your case a custom formatter for JSON is the right answer. 

You don't have pre-built JSON, you have a string. Grape can't guess that it's JSON. And the JSON formatter will translate a string into ... JSON.

**Comment by iabdulin** [comment]
My issue is that even if I explicitly set a :txt or :binary formatter, it still applies :json formatter when I set 
`content_type 'application/json'`
or
`header 'Content-Type', 'application/json'`.

Setting a content_type or header should not overwrite formatter.

**Comment by dblock** [comment]
You're sending JSON, not binary. So you should just be replacing the default JSON formatter with yours with `formatter :json,-> ` as you did above. Does that not work?

**Comment by iabdulin** [comment]
It works, I just think that there should be a way to specify 'application/json' header without grape applying :json formatter.
I couldn't find a way to send this header.

`env['api.format'] = :txt` — sends empty content-type
adding `content_type 'application/json'` or `header 'Content-Type', 'application/json'` invokes :json formatter

**Comment by iabdulin** [comment]
Even if I do this:
```ruby
env['api.format'] = :txt
content_type 'application/json'
body {key: 123}.to_json
```
grape still applies :json formatter

**Comment by dblock** [comment]
This should work as is, and we have specs for this. I think something else is going on.

Turn it onto a spec or a small repro of some kind? Maybe add an example to https://github.com/ruby-grape/grape-on-rack?

**Comment by waynerobinson** [comment]
I'm having exactly the same issue.

If this is used:

```ruby
env['api.format'] = :binary # or :txt
content_type 'application/json' # or header 'Content-Type', 'application/json'
{key: 123}.to_json
```

Then it still calls the JSON formatter. If the `content_type` method is left off, it doesn't.

**Comment by waynerobinson** [comment]
Here is an example using grape-on-rack:

https://github.com/waynerobinson/grape-on-rack

Please take a look at the extra methods added to `Acme::Ping`.

The only one that doesn't call the JSON formatter is the one where the content type is set to `text/plain`.

**Comment by dblock** [comment]
This does look like a bug, and shouldn't have been closed. Let's fix it in #2171.
